### PR TITLE
feat: entity view for adminmenu

### PR DESCRIPTION
**Pull request by Spudgun**
**Describe Pull request**
I believe this PR will be beneficial to the wider community. It's an expansion on the developer menu and enables the user to view a whole host of properties related to entities. 

1. Four different modes of viewing entities, which can be combined or used separately `Freeaim, Vehicles, Ped or Object Entity View`
2. A scale to set the view distance for the `Vehicle, Ped and Object Entity View`
3. The ability to toggle Freezing of entities whilst using `Vehicle, Ped or Object Entity View`
4. Delete lazer functionality added to the `Freeaim Entity View` mode. 

**Note:** I have limited the scale to 50 units. Even then, I believe 50 is too high. You'd most probably want to be utilizing around 15-20 at most. However, the slider is added in so the user can decide that themselves.

*See Below for screenshot examples*

**New Menu Option in Developer Options**
1Vjrivf

**Entity View Mode Menu**
9v17jtM

**Freeaim mode - Used a lot of the delete laser code**
*I have also added functionality for this to be used in no-clip. This is done by changing which camera the raytrace is using. If in noclip, it will use the camera created. Otherwise, it'll use the Gameplay Camera.*
8bf60373-cb9a-40b8-b785-ecd69377ed08

**View objects in the world**
EnbNr00

**View vehicles in the world**
tlt5OWj

**view peds information in the world**
b4dd63f9-4550-430a-a226-d4aabbfa515a

**View objects, vehicles and peds in the world**
lZxkaYj

**Questions (please complete the following information):**
- Have you personally loaded this code into an updated qbcore project and checked all it's functionality? Yes
- Does your code fit the style guidelines? yes
- Does your PR fit the contribution guidelines? yes

### CODE REVIEW

**Comment by Spudgun** [comment]
I also just added an object name display. 

This is done by adding in a file called `entity.lua` in a new folder called `entityhashes`. It has a table with around 54000 known entity hashes and their respective name. Each entity name is stored with its hash as the key value so it's super quick to index.

**Entity object name**
image

**Comment by Spudgun** [comment]
The above commit adds the ability to copy some information from the `Freeaim Entity` such as `Object name`, `Object hash`, `Object Heading`, `Object Coords`, and `Object Rotation` to the clipboard.

**Copy Freeaim Entity Info to Clipboard Menu Item**
image

File: client/client.lua [hunk 1]
```diff
@@ -344,48 +347,136 @@ local vehicledev_button = menu11:AddButton({
...
-})
 local noclip_button = menu11:AddCheckbox({
     icon = '🎥',
     label = Lang:t("menu.noclip"),
-    value = menu11,
+    value = nil,
     description = Lang:t("desc.noclip_desc")
 })
 
+local entity_view_button = menu11:AddButton({
```
Comment (restray) [inline client/client.lua:361]: This variable is unused! What is the usage of it?

File: client/client.lua [hunk 2]
```diff
@@ -437,6 +527,29 @@ local function CopyToClipboard(dataType)
             string = h
         })
         QBCore.Functions.Notify(Lang:t("success.heading_copied"), "success")
+    elseif dataType == 'freeaimEntity' then
+        local entity = GetFreeAimEntity()
+
+        if entity then
+            local entityHash    = GetEntityModel(entity)
+            local entityName    = Entities[entityHash] ~= nil and Entities[entityHash] or "Unknown"
```
Comment (restray) [inline client/client.lua:535]:
```suggestion
local entityName    = Entities[entityHash] or "Unknown"
```
No need of a "non nil" condition for this!

File: client/entity_view.lua [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    EntityViewEnabled = true
+    Citizen.CreateThread(function()
+        while EntityViewEnabled do         
+            Citizen.Wait(0)
+            local freezeDesc = (EntityObjectView or EntityPedView or EntityVehicleView)
+            DrawTitle("~y~ENTITY VIEW~w~\n"..(EntityFreeAim and "\n[~y~E~w~] - Delete Entity~w~" or "").. (freezeDesc and '\n[~y~G~w~] - Freeze All Entities' or ''), 0.15, freezeDesc and 0.14 or 0.11)
+
+            local entity        = nil
+            local playerPed     = PlayerPedId()
+            local playerCoords  = GetEntityCoords(playerPed)
```
Comment (restray) [inline client/entity_view.lua:379]: Watch for empty space at the EOL

File: client/entity_view.lua [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+            end
+
+            if IsControlJustReleased(0, 38) then -- Delete entity
+                -- Set as missionEntity so the object can be remove (Even map objects)
+			    SetEntityAsMissionEntity(entity, true, true)
+			    DeleteEntity(entity)
+
+                if not DoesEntityExist(entiy) then
+                    QBCore.Functions.Notify('Entity Deleted', 'success')
+                end
```
Comment (restray) [inline client/entity_view.lua:415]: Watch for empty space at the EOL

File: client/entity_view.lua [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+            end
+            
+            if EntityVehicleView then
+                GetVehicle(playerPed, playerCoords)
+            end
+
+            if EntityFreeAim then
+                local color = {r = 255, g = 200, b = 0, a = 200}
+                local position = GetEntityCoords(PlayerPedId())
+                hit, coords, entity = RayCastGamePlayCamera(1000.0)
```
Comment (restray) [inline client/entity_view.lua:396]: Should this be global ? Please set local, don't need to expose those

File: client/entity_view.lua [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+                if not DoesEntityExist(entiy) then
+                    QBCore.Functions.Notify('Entity Deleted', 'success')
+                end          
+            end
+
+            if EntityPedView or EntityVehicleView or EntityObjectView then
+                if IsControlJustReleased(0, 47) then -- Freeze entities
+                    FreezeEntities = not FreezeEntities
+
+                    QBCore.Functions.Notify('Freeze all entities now '.. (FreezeEntities and "enabled" or "disabled"), FreezeEntities and 'success' or 'error')
```
Comment (restray) [inline client/entity_view.lua:422]: Watch for empty space at the EOL

**Review by restray — CHANGES_REQUESTED** [review]
Some linting errors, but huge modification!
If you enable Ped & Vehicle view mode in the same time, this will make the UI impossible to read!
A little bug on entity freeze (or I don't know how to use it): if I press on G, Ped continue to drive, walk, etc...

Thanks for your work!

File: client/client.lua (hunk 1 above)
Comment (Spudgun) [reply client/client.lua:361]: This is used for setting the Initial menu under the `Developer Options Menu` which links to `menu15`. Which has all the settings for the new `Entity View Mode`.

**Comment by Spudgun** [comment]
> Thanks for your work!

Figured this would just be a good tool for developers for viewing entity information and could help out with development of servers. So happy to do the work as it'll help everyone! 😊

> Some linting errors, but huge modification! If you enable Ped & Vehicle view mode in the same time, this will make the UI impossible to read! 

Yeah, I have noticed this.. I'm not 100% sure on how to fix this, other than maybe drawing a bounding box on an entity if you're say x amount of distance away and only then drawing the entity information? Maybe that would be a better implementation? You still get UI feedback of where entities are but only the entity information if you're close to them.

> A little bug on entity freeze (or I don't know how to use it): if I press on G, Ped continue to drive, walk, etc...

In regards to the freezing - This is only implemented for `View Peds`, `View Vehicles`, and `View Objects`. It is toggled as either `on` or `off` and will set `FreezeEntityPosition` to either `true` or `false` for that entity based on whether or not you have toggled it on or off and if you're within the `View Distance` you've set. 

Maybe a better implementation of this will adding the entities to a table and freezing/unfreezing them based on your key input. That way you can unfreeze entities if you're not near them. It would also mean we can use the `Freeaim Entity View` to freeze an object as we'd just add that to a table.

Happy to hear any opinions on how you'd like that to be implemented though.

**Comment by Spudgun** [comment]
Alright, I have made the change to how much entity information is actually displayed on the screen when `vehicle`, `ped` or `object` entities in the world are displayed.

Information will only display if you are within `5` units of the object. Otherwise, it'll just draw a bounding box around the entity. See the below screenshot for reference.

image

I also rewrote the logic for the static information boxes. `DrawRect` should now be dynamic to the values it gets given. i.e. users should be able to move that around by defining only the `x` and `y` of where they want the Information to be displayed.

**Comment by Sna-aaa** [comment]
Just tested it now, i've no errors at all, this works nicely for me
And a really nice and usefull tool thanks a lot for your work

**Comment by Spudgun** [comment]
> Just tested it now, i've no errors at all, this works nicely for me And a really nice and usefull tool thanks a lot for your work

Glad you like it! Always good to expand on dev tools where possible 😊

**Comment by Spudgun** [comment]
Alright, I believe this PR just needs to be reviewed now.

Thanks for all the input 😊

**Comment by Spudgun** [comment]
@GhzGarage Did you have a chance to check this out further? I know you're probably insanely busy with other PR for the framework but figured I'd just bump this

**Comment by GhzGarage** [comment]
> @GhzGarage Did you have a chance to check this out further? I know you're probably insanely busy with other PR for the framework but figured I'd just bump this

Just add that one line of code to the en.lua and i'll test this tonight and hopefully get it pushed through which doesn't seem like it would be an issue based on other comments

**Pull request merged**